include/boost/math/special_functions/jacobi_theta.hpp

```diff
@@ -0,0 +1,672 @@
+                "tau must be greater than 0 but got %1%.", tau, pol);
+
+    if (abs(z) == 0.0)
+        return result;
+
+    if (tau < 1.0) {
+        z = fmod(z, constants::two_pi<RealType>());
+        while (z > constants::pi<RealType>()) {
+            z -= constants::two_pi<RealType>();
+        }
+        while (z < -constants::pi<RealType>()) {
+            z += constants::two_pi<RealType>();
+        }
+
+        return _IMAGINARY_jacobi_theta1tau(z, 1/tau, pol);
+    }
+
+    do {
+        q_n = exp(-tau * constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5) );
+        delta = q_n * sin(RealType(2*n+1)*z);
+        if (n%2)
+            delta = -delta;
+
+        result += RealType(2) * delta;
+        n++;
+    } while (abs(q_n) > eps || (abs(result) > eps * eps && abs(q_n/result) > eps));
+
+    if (abs(result) < eps * eps)
+        return RealType(0);
+
+    return result;
+}
+
+template <class T>
+inline T jacobi_theta1tau(T z, T tau) {
+    return jacobi_theta1tau(z, tau, policies::policy<>());
+}
+
+// First Jacobi theta function (Parameterized by q)
+// = 2 * Σ (-1)^n * q^(n+1/2)^2 * sin((2n+1)z)
+template <class RealType, class Policy>
+inline RealType
+jacobi_theta1(RealType z, RealType q, const Policy& pol) {
+    BOOST_MATH_STD_USING
+    if (q <= 0.0 || q >= 1.0) {
+        return policies::raise_domain_error<RealType>("boost::math::jacobi_theta1<%1%>(%1%)",
+                "q must be greater than 0 and less than 1 but got %1%.", q, pol);
+    }
+    return jacobi_theta1tau(z, -log(q)/constants::pi<RealType>(), pol);
+}
+
+template <class T>
+inline T jacobi_theta1(T z, T q) {
+    return jacobi_theta1(z, q, policies::policy<>());
+}
+
+// Second Jacobi theta function (Parameterized by tau - assumed imaginary)
+// = 2 * Σ exp(iπτ*(n+1/2)^2) * cos((2n+1)z)
+template <class RealType, class Policy>
+inline RealType
+jacobi_theta2tau(RealType z, RealType tau, const Policy& pol)
+{
+    BOOST_MATH_STD_USING
+    unsigned n = 0;
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType q_n, delta, result = RealType(0);
+
+    if (tau <= 0.0) {
+        return policies::raise_domain_error<RealType>("boost::math::jacobi_theta2tau<%1%>(%1%)",
+                "tau must be greater than 0 but got %1%.", tau, pol);
+    } else if (tau < 1.0 && abs(z) == 0.0) {
+        return jacobi_theta4tau(z, 1/tau, pol) / sqrt(tau);
+    } else if (tau < 1.0) { // DLMF 20.7.31
+        z = fmod(z, constants::two_pi<RealType>());
+        while (z > constants::pi<RealType>()) {
+            z -= constants::two_pi<RealType>();
+        }
+        while (z < -constants::pi<RealType>()) {
+            z += constants::two_pi<RealType>();
+        }
+
+        return _IMAGINARY_jacobi_theta4tau(z, 1/tau, pol);
+    }
+
+    do {
+        q_n = exp(-tau * constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5));
+        delta = q_n * cos(RealType(2*n+1)*z);
+        result += RealType(2) * delta;
```

Comment:
I worry that this cast of `2` to `RealType(2)` will prevent generation of an exponent shift in the exponent and for a mul. I'll see what godbolt says in a bit . .

Reply:
Easy enough to change!

Reply:
+1, also in the multiprecision case, multiplying by a literal integer is *much* more efficient than the cast.  BTW we assume in our conceptual requirements that all number types will interact correctly with integer types, so multiplying/dividing by literals (mostly 2!) is pretty commonplace in the library.

Reply:
Fixed in latest commit. I opted for `delta + delta` which I imagine compiles to the same thing.